python/pyiceberg/expressions/visitors.py

```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
+        idx = term.ref().field.field_id
+
+        if self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_less_than(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.lower_bounds:
+            lower = from_bytes(field.field_type, self.lower_bounds[idx])
+
+            if self._is_nan(lower):
+                # NaN indicates unreliable bounds. See the InclusiveMetricsEvaluator docs for more.
+                return ROWS_MIGHT_MATCH
+
+            if lower >= literal.value:
+                return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_less_than_or_equal(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.lower_bounds:
+            lower = from_bytes(field.field_type, self.lower_bounds[idx])
+
+            if self._is_nan(lower):
+                # NaN indicates unreliable bounds. See the InclusiveMetricsEvaluator docs for more.
+                return ROWS_MIGHT_MATCH
+
+            if lower > literal.value:
+                return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_greater_than(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.upper_bounds and from_bytes(field.field_type, self.upper_bounds[idx]) <= literal.value:
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_greater_than_or_equal(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.upper_bounds and from_bytes(field.field_type, self.upper_bounds[idx]) < literal.value:
+            return ROWS_CANNOT_MATCH
```

Comment:
Why do the greater-than implementations omit the NaN check that was used in less-than and equal implementations?

Reply:
Oh, I see Java does it too. My guess is that we don't expect NaN for upper bounds, but it's probably safer just to run the check everywhere.

Reply:
@Fokko, looks like this wasn't updated. The less_than and greater_than cases still differ in structure because the null check isn't there.

Reply:
I misinterpreted this comment, let me add those checks for each of the expressions.

Reply:
I can add those to Java as well.